### ISSUE TITLE
Scope channel names

### ISSUE THREAD

**Issue by chrylis** [spring-cloud/spring-cloud-stream#949]
_Note: This request may be a breaking change._

On any binding interface, SCS uses the channel name for the bean name regardless of whether `@Input` or `@Output` is applied. This can cause problems when both a sink and a source to the same channel are in the same application (e.g., the channel is used for fan-out, but the application itself does want to process event messages), since the bean definition by name is overridden in the context.

I suggest scoping the channel names with something like `"input.CHANNEL"` to produce the bean names.

**Comment by markfisher** [comment]
Can you clarify what you mean be "the channel is used for fan-out, but the application itself does want to process event messages"?

**Comment by chrylis** [comment]
Application Foo has a message-driven architecture. When a user Does The Thing, a `UserDidTheThing` message is sent to the MQ on `channel-user-did-thing`. Application Foo itself has an `@StreamListener` that listens for messages on that channel to keep on with business logic, while application Analytics and application Logging also listen to that channel and make their own entries.

The difficulty is that if application Foo has both source and sink interfaces for `channel-user-did-thing`, then both of them try to use the same bean name in the application context, which can cause problems with things like qualifiers.

**Comment by mbogoevici** [comment]
>The difficulty is that if application Foo has both source and sink interfaces for channel-user-did-thing, then both of them try to use the same bean name in the application context, which can cause problems with things like qualifiers.

You shouldn't need to use the same name for the channel. You can have two separate channels with different names (e.g. `fooInput` and `fooOutput` and map them to the same destination via `s.c.s.binding.fooInput.destination=foo` and `s.c.s.binding.fooOutput.destination=foo`.

**Comment by chrylis** [comment]
That then requires explicitly mapping all of the channels, which mostly defeats the purpose of using SCS over skipping it and writing the whole configuration in Spring Integration directly.

**Comment by artembilan** [comment]
@chrylis ,

Describe, please, how you can have different beans with the same name?

The SCSt is firstly a Spring Application and only after that it provides some goodies to abstract from the hard configuration a bit.

**Comment by mbogoevici** [comment]
>That then requires explicitly mapping all of the channels, which mostly defeats the purpose of using SCS over skipping it and writing the whole configuration in Spring Integration directly.

I don't think so, as in Spring Integration you would still have to declare separate beans (with separate names) for the channels. Spring Cloud Stream adds more automation to the process, as well as a standardized way to do configuration, content negociation, etc. while allowing you to use Spring Integration components (as well as other features) in your application. 

However, should you feel that using inbound/outbound channel adapters directly is better suited for your use case, by all means do so.

**Comment by chrylis** [comment]
> Describe, please, how you can have different beans with the same name?

You can't; that's the problem. For example, I have an application that is responsible for sending mail in response to RabbitMQ messages. Those messages may come from a different application on the same MQ or from another component in the same application in response to some other event.

If the name of this queue in Rabbit is `send-mail`, I can declare an interface that looks like this:

```
public interface MailSender {
    String CHANNEL = "send-mail";

    @Output(CHANNEL)
    MessageChannel outChannel();
```

SCS will correctly bind RabbitMQ here, creating a channel and registering the bean with the name `send-mail`. However, if I _also_ want to listen on that queue in my service application, but not in every application, I declare a companion interface inside the application with the actual mail code:

```
public interface MailListener {

    @Input(MailSender.CHANNEL)
    SubscribableChannel inChannel();
}
```

At this point, SCS creates a DirectChannel, reusing the bean name `send-mail` for it, and the bean name gets overridden in the context. (It's still injectable by type as long as it doesn't need to be disambiguated.)

> However, should you feel that using inbound/outbound channel adapters directly is better suited for your use case, by all means do so.

Do you mean using the SI configuration directly? I'm considering this, and I'll go that path if I can't get the pieces to snap together, but this _seems_ like a common enough use case that there should be a workable setup.

**Comment by artembilan** [comment]
The problem here that place everything upside-down. You trap yourself overriding existing beans! 😄 

The ` @Input("foo")` means *entrance* to the room.
The `   @Output("foo")` - an *exit*, of course.

And hijacking one of the with the name of another one just make a hole in pitch black without any usability!

The room is something like this:
```
IntgrationFlows.from("foo")
     .split(...)
     .handle(...)
     .transform(...)
     .aggregate(...)
     .channel("foo")
```

So, one more time: it is Spring first of all and everything are beans. Reusing bean names just bring us bean overriding.

That's how things work here. Nothing more. Really: no magic!

**Comment by chrylis** [comment]
I understand that about _bean_ names. The catch is that the basic convenient SCS setup uses the exact same name for the bean and the queue binding.

You can get around that by explicitly providing a mapping, but that starts getting into manual configuration, and specifically in Boot there's no simple way to pull in multiple fragmentary YAML files to provide a "library" of channel bindings (you could add a profile per module, but that would get complicated fast).

> And hijacking one of the with the name of another one just make a hole in pitch black without any usability!

What I was _trying_ to make happen was something like `tee`!

**Comment by chrylis** [comment]
I think that part of the issue is that my mental model of `@Input(FOO)` (or output) is of FOO (a "channel") as a RabbitMQ queue instead of as a Spring adapter bean. The fact that SCS binds them to the same name by default is convenient, but it makes the "loopback" use case tricky, and there's no way that I've been able to figure to provide reusable binding defaults (e.g., no way to say `send-mail.input` == `send-mail.output` == queue named `send-mail` in something that I can cleanly pull in via autoconfiguration).

Maybe what I'm actually looking for is to be able to say something like

```
@Input(channel = "send-mail.input", defaultBinding = "send-mail")
```

```
@Output(channel = "send-mail.output", defaultBinding = "send-mail")
```

If that makes more sense and seems like a reasonable way to accomplish what I'm looking to do, I'll call this feature request an XY problem and open a different one.

**Comment by mbogoevici** [comment]
Hi @chrylis,

Thanks for the suggestion! This is a much better solution, as we might want to be able to set other default binding properties there (e.g. `contentType` for an outbound message, possibly even partitioning properties). In any case, please open a new issue and we'll consider it for our next major release.
